client/components/combobox.jsx

```diff
@@ -1,128 +1,167 @@
-const React = require('react');
-const createClass = require('create-react-class');
-const _ = require('lodash');
 require('./combobox.less');
+import React, { useState, useRef, useEffect } from 'react';
+const _ = require('lodash');
 
-const Combobox = createClass({
-	displayName     : 'Combobox',
-	getDefaultProps : function() {
-		return {
-			className   : '',
-			trigger     : 'hover',
-			default     : '',
-			placeholder : '',
-			autoSuggest : {
-				clearAutoSuggestOnClick : true,
-				suggestMethod           : 'includes',
-				filterOn                : []  // should allow as array to filter on multiple attributes, or even custom filter
-			},
-		};
-	},
-	getInitialState : function() {
-		return {
-			showDropdown : false,
-			value        : '',
-			options      : [...this.props.options],
-			inputFocused : false
+const Combobox = ({ onSelect, onEntry, autoSuggest = { filterOn: ['data-value'], suggestMethod: 'includes', clearAutoSuggestOnClick: false }, ...props })=>{
+	const [inputValue, setInputValue] = useState(props.value || '');
+	const [showDropdown, setShowDropdown] = useState(false);
+	const [inputFocused, setInputFocused] = useState(false);
+	const [currentOption, setCurrentOption] = useState(-1);
+	const [filteredOptions, setFilteredOptions] = useState(React.Children.toArray(props.children));
+	const inputRef = useRef(null);
+	const optionRefs = useRef([]);
+	const componentRef = useRef(null);
+
+	useEffect(()=>{
+		const handleClickOutside = (evt)=>{
+			if(showDropdown && componentRef.current && !componentRef.current.contains(evt.target)) {
+				setShowDropdown(false);
+			}
 		};
-	},
-	componentDidMount : function() {
-		if(this.props.trigger == 'click')
-			document.addEventListener('click', this.handleClickOutside);
-		this.setState({
-			value : this.props.default
-		});
-	},
-	componentWillUnmount : function() {
-		if(this.props.trigger == 'click')
-			document.removeEventListener('click', this.handleClickOutside);
-	},
-	handleClickOutside : function(e){
-		// Close dropdown when clicked outside
-		if(this.refs.dropdown && !this.refs.dropdown.contains(e.target)) {
-			this.handleDropdown(false);
+
+		document.addEventListener('pointerdown', handleClickOutside);
```

Comment:
This is also adding a new event listener every time `showDropdown` changes.

If you want to add a listener once when the component first mounts, change the `[showDropdown]` to an empty array `[]`

Reply:
If just change this to onMount `[]`, it doesn't work.  But it works if i remove the `showDropdown` condition in the handler:

```
const handleClickOutside = (evt)=>{
		console.log(componentRef.current, showDropdown)
		if(componentRef.current && !componentRef.current.contains(evt.target)) {
			setShowDropdown(false);
		}
	};
```

So that's what i'll do.

Reply:
Without any dependency array it now adds a new listener on *every* render, which is even worse.

Let's see if we can figure out why `[]` didn't work